### PR TITLE
Decompress limits

### PR DESCRIPTION
**Pull request by Corion**
This supercedes https://github.com/libwww-perl/HTTP-Message/pull/42

### CODE REVIEW

File: lib/HTTP/Message.pm [hunk 1]
```diff
@@ -288,34 +301,86 @@ sub decoded_content
...
+			my $input = "$$content_ref";
+			$content_ref = \$input;
+			$content_ref_iscopy++;
+		    };
+		    my ($i, $status) = Compress::Raw::Zlib::Inflate->new(
+			%limiter_options,
+			ConsumeInput => 0, # overridden by Zlib if we have %limiter_options :-(
+			WindowBits => Compress::Raw::Zlib::WANT_GZIP_OR_ZLIB(),
+		    );
+		    #warn "Foo: $status";
```
Comment (simbabque) [inline lib/HTTP/Message.pm:333]: There's a warning left in here

File: lib/HTTP/Message.pm [hunk 2]
```diff
@@ -288,34 +301,86 @@ sub decoded_content
...
+
 		    $@ and die "Can't unbrotli content: $@";
 		    $content_ref = \$output;
 		    $content_ref_iscopy++;
 		}
 		elsif ($ce eq "x-bzip2" or $ce eq "bzip2") {
-		    require IO::Uncompress::Bunzip2;
+		    require Compress::Raw::Bzip2;
+
+		    if( ! $content_ref_iscopy ) { #and keys %limiter_options) {
```
Comment (simbabque) [inline lib/HTTP/Message.pm:361]: Is this comment left over code or does it mean something?

File: t/message-decode-brotlibomb.t [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+# Manual decompression check
+my $output = $compressed;
+for( 1 ) {
+    my $unbro = IO::Uncompress::Brotli->create();
+    $output = $unbro->decompress($compressed);
+};
+
+$headers = HTTP::Headers->new(
+    Content_Type => "application/xml",
+    Content_Encoding => 'br' # say my name three times
```
Comment (simbabque) [inline t/message-decode-brotlibomb.t:57]: It's only there one time

**Review by simbabque — CHANGES_REQUESTED** [review]
This is very impressive work. Thank you for that. Could you please remove those leftover commented out debug messages?

I think we should also check why the Brotli test skips in CI.

File: lib/HTTP/Message.pm (hunk 2 above)
Comment (Corion) [reply lib/HTTP/Message.pm]: Addressed by [58e5573](https://github.com/libwww-perl/HTTP-Message/pull/181/commits/58e5573aac0d74a3c84339ffd56df868afc10e1f)

File: t/message-decode-brotlibomb.t (hunk 3 above)
Comment (Corion) [reply t/message-decode-brotlibomb.t]: Addressed by [58e5573](https://github.com/libwww-perl/HTTP-Message/pull/181/commits/58e5573aac0d74a3c84339ffd56df868afc10e1f)

File: lib/HTTP/Message.pm (hunk 1 above)
Comment (Corion) [reply lib/HTTP/Message.pm]: Addressed by [58e5573](https://github.com/libwww-perl/HTTP-Message/pull/181/commits/58e5573aac0d74a3c84339ffd56df868afc10e1f)

**Comment by Corion** [comment]
The Brotli test skips in CI because `cpanm` is run from the action without `--with-recommends` . I don't know enough about the Dockerfile setup and the Github actions used to know how to coax this into actually installing things.

**Comment by oalders** [comment]
> The Brotli test skips in CI because cpanm is run from the action without --with-recommends . I don't know enough about the Dockerfile setup and the Github actions used to know how to coax this into actually installing things.

That also skips in `master`, so we don't need to address it here.

File: t/message-decode-brotlibomb.t [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+use strict;
+use warnings;
+
+use Test::More;
+
+use HTTP::Headers    qw();
+use HTTP::Response   qw();
+
+my $ok = eval {
```
Comment (oalders) [inline t/message-decode-brotlibomb.t:11]: I think we could use `Test::Needs` here. That would be a little less verbose.

**Review by oalders — CHANGES_REQUESTED** [review]

File: t/message-decode-brotlibomb.t (hunk 4 above)
Comment (Corion) [reply t/message-decode-brotlibomb.t]: I've moved the manual checking/loading of modules into `Test::Needs` (which already is a prerequisite) for this and for `t/message-decode-zipbomb.t` as well.

**Comment by oalders** [comment]
I will probably fix the merge conflict and look at merging this tomorrow, so if anyone has objections, now would be a good time to bring them up.

**Comment by oalders** [comment]
Thanks so much for revisiting this, @Corion! I've rebased and merged at the command line. New release will be on CPAN shortly.

**Pull request closed without merging**